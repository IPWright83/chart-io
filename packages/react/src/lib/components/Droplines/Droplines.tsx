import * as d3 from "@d3-chart/d3";
import type { IDropline } from "@d3-chart/types";
import { useEffect } from "react";
import { useSelector } from "react-redux";

import { chartSelectors, eventSelectors, IState } from "../../store";

export interface IDroplinesBaseProps {
    /**
     * The layer to be rendered upon. Typically this is an `<svg:g>` or a fake HTMLElement when using canvas.
     * @default undefined
     */
    layer: React.MutableRefObject<Element>;

    /**
     * Should horizontal droplines be shown?
     * @default true
     */
    showHorizontal?: boolean;

    /**
     * Should vertical droplines be shown?
     * @default true
     */
    showVertical?: boolean;
}

/**
 * This component renders the droplines that are triggered from various plots
 * @return {ReactElement}  The Droplines component
 */
export function Droplines({ layer, showVertical = true, showHorizontal = true }: IDroplinesBaseProps) {
    const animationDuration = useSelector((s: IState) => chartSelectors.animationDuration(s));
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const droplines = useSelector((s: IState) => eventSelectors.droplines(s)).filter(
        (dl) => (dl.isVertical && showVertical) || (dl.isHorizontal && showHorizontal)
    );

    useEffect(() => {
        if (!layer.current) return;

        const join = d3
            .select(layer.current)
            .selectAll(".dropline")
            .data(droplines, (d: IDropline) => `${d.x1}-${d.x2}-${d.y1}-${d.y2}`);

        join.exit().remove();

        // Add any new droplines
        join.enter()
            .append("line")
            .attr("class", "d3-chart dropline")
            .attr("point-events", "none")
            .style("stroke-dasharray", theme.droplines.strokeDasharray)
            .style("stroke-opacity", theme.droplines.strokeOpacity)
            .style("stroke-width", theme.droplines.strokeWidth)
            .style("stroke", (d) => `${d.color}`)
            .attr("x1", (d) => d.x1)
            .attr("x2", (d) => d.x1)
            .attr("y1", (d) => d.y1)
            .attr("y2", (d) => d.y1)
            .transition()
            .delay(animationDuration)
            .duration(animationDuration)
            .attr("x2", (d) => d.x2)
            .attr("y2", (d) => d.y2);
    }, [animationDuration, layer, droplines]);

    return null;
}
