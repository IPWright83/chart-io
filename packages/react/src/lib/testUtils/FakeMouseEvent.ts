export interface IFakeMouseEventValues {
    pageX?: number;
    pageY?: number;
    offsetX?: number;
    offsetY?: number;
    x?: number;
    y?: number;
}

/**
 * Create a Fake MouseEvent for testing
 */
export class FakeMouseEvent extends MouseEvent {
    /**
     * Construct a new instance of a FakeMoustEvent
     * @param  type        The type of event, e.g. "click", "mouseenter"
     * @param  values      An object containing values for the x,y co-ordinates for the event
     */
    constructor(type: string, values: IFakeMouseEventValues) {
        const { pageX, pageY, offsetX, offsetY, x, y, ...mouseValues } = values;
        super(type, mouseValues);

        Object.assign(this, {
            offsetX: offsetX || 0,
            offsetY: offsetY || 0,
            pageX: pageX || 0,
            pageY: pageY || 0,
            x: x || 0,
            y: y || 0,
        });
    }
}
