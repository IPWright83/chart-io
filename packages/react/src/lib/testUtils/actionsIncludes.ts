/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
