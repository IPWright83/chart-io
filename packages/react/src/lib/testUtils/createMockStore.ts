import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
