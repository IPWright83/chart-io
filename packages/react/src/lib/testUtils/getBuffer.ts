/**
 * Returns the image buffer for a Canvas
 * @param canvas    The HTML canvas that we're going to extract the buffer from
 * @returns         The image buffer
 */
export function getBuffer(canvas: HTMLCanvasElement) {
    const image = canvas.toDataURL("image/png");
    const data = image.replace(/^data:image\/\w+;base64,/, "");
    // eslint-disable-next-line no-undef
    return Buffer.from(data, "base64");
}
