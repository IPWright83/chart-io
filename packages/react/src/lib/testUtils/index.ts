export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
