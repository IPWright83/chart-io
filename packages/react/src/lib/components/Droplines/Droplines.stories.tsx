import isChromatic from "chromatic/isChromatic";
import React from "react";
import { Provider } from "react-redux";

import { themes } from "../../themes";
import { Droplines } from ".";

import { createMockStorybookStore } from "../../testUtils";

export default {
    title: "Components/Droplines",
    component: Droplines,
    parameters: {
        chromatic: { delay: 300 },
    },
};

const DroplinesTemplate = (args) => {
    const store = createMockStorybookStore({
        chart: {
            animationDuration: isChromatic() ? 0 : 1000,
            theme: themes.light,
        },
        event: {
            droplines: [
                { isHorizontal: true, color: "steelblue", x1: 150, x2: 0, y1: 50, y2: 50 },
                { isVertical: true, color: "steelblue", x1: 150, x2: 150, y1: 50, y2: 150 },
            ],
        },
    });

    return (
        <Provider store={store}>
            <svg>
                <Droplines showVertical={args.showVertical} showHorizontal={args.showHorizontal} />
            </svg>
        </Provider>
    );
};

export const Default = DroplinesTemplate.bind({});
Default.storyName = "Horizontal & Vertical Droplines";
Default.args = {
    showVertical: true,
    showHorizontal: true,
};

export const VerticalDroplines = DroplinesTemplate.bind({});
VerticalDroplines.storyName = "Vertical Droplines";
VerticalDroplines.args = {
    showVertical: true,
    showHorizontal: false,
};

export const HorizontalDroplines = DroplinesTemplate.bind({});
HorizontalDroplines.storyName = "Horizontal Droplines";
HorizontalDroplines.args = {
    showVertical: false,
    showHorizontal: true,
};
