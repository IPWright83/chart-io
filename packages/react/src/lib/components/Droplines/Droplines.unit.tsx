import { themes } from "@chart-io/core";

import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "../../testUtils";

import { Droplines } from ".";
import { Droplines as DroplinesBase } from "./Droplines";

describe("Droplines", () => {
    const store = createMockStore({
        chart: {
            theme: themes.light,
        },
        event: {
            droplines: [
                { isHorizontal: true, color: "red", x1: 50, x2: 0, y1: 50, y2: 50 },
                { isVertical: true, color: "red", x1: 50, x2: 50, y1: 50, y2: 0 },
            ],
        },
    });

    it("should render correctly", async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render just vertical lines correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines showHorizontal={false} />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render just horizontal lines correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines showVertical={false} />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };

        render(
            <Provider store={store}>
                <svg>
                    <DroplinesBase layer={layer} />
                </svg>
            </Provider>,
        );

        // Should be empty
        expect(layer).toMatchSnapshot();
    });
});
