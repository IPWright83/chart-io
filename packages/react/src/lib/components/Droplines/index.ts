import { withSVG } from "../../hoc";
import { Droplines as DroplinesBase, IDroplinesBaseProps } from "./Droplines";

export interface IDroplinesProps extends Omit<IDroplinesBaseProps, "layer"> {}

export const Droplines = withSVG<IDroplinesProps>(DroplinesBase, "droplines");
