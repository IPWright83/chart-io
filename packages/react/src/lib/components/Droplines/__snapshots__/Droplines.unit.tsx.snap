// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Droplines should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-io dropline"
        point-events="none"
        style="stroke-dasharray: 4; stroke-opacity: 1; stroke-width: 2; stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
      <line
        class="chart-io dropline"
        point-events="none"
        style="stroke-dasharray: 4; stroke-opacity: 1; stroke-width: 2; stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should render just horizontal lines correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-io dropline"
        point-events="none"
        style="stroke-dasharray: 4; stroke-opacity: 1; stroke-width: 2; stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should render just vertical lines correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-io dropline"
        point-events="none"
        style="stroke-dasharray: 4; stroke-opacity: 1; stroke-width: 2; stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should skip if there is no layer avaliable 1`] = `
{
  "current": null,
}
`;
